examples/mixture/plot_gmm_selection.py

```diff
@@ -3,108 +3,161 @@
+
+from sklearn.mixture import GaussianMixture
+from sklearn.model_selection import GridSearchCV
+
+
+def gmm_bic_score(estimator, X):
+    """Callable to pass to GridSearchCV that will use the BIC score."""
+    # Make it negative since GridSearchCV expects a score to maximize
+    return -estimator.bic(X)
 
-# Generate random sample, two components
-np.random.seed(0)
-C = np.array([[0.0, -0.1], [1.7, 0.4]])
-X = np.r_[
-    np.dot(np.random.randn(n_samples, 2), C),
-    0.7 * np.random.randn(n_samples, 2) + np.array([-6, 3]),
-]
 
-lowest_bic = np.infty
-bic = []
-n_components_range = range(1, 7)
-cv_types = ["spherical", "tied", "diag", "full"]
-for cv_type in cv_types:
-    for n_components in n_components_range:
-        # Fit a Gaussian mixture with EM
-        gmm = mixture.GaussianMixture(
-            n_components=n_components, covariance_type=cv_type
-        )
-        gmm.fit(X)
-        bic.append(gmm.bic(X))
-        if bic[-1] < lowest_bic:
-            lowest_bic = bic[-1]
-            best_gmm = gmm
-
-bic = np.array(bic)
-color_iter = itertools.cycle(["navy", "turquoise", "cornflowerblue", "darkorange"])
-clf = best_gmm
-bars = []
+param_grid = {
+    "n_components": range(1, 7),
+    "covariance_type": ["spherical", "tied", "diag", "full"],
+}
+grid_search = GridSearchCV(
+    GaussianMixture(), param_grid=param_grid, scoring=gmm_bic_score
+)
+grid_search.fit(X)
 
+# %%
 # Plot the BIC scores
-plt.figure(figsize=(8, 6))
-spl = plt.subplot(2, 1, 1)
-for i, (cv_type, color) in enumerate(zip(cv_types, color_iter)):
-    xpos = np.array(n_components_range) + 0.2 * (i - 2)
-    bars.append(
-        plt.bar(
-            xpos,
-            bic[i * len(n_components_range) : (i + 1) * len(n_components_range)],
-            width=0.2,
-            color=color,
-        )
-    )
-plt.xticks(n_components_range)
-plt.ylim([bic.min() * 1.01 - 0.01 * bic.max(), bic.max()])
-plt.title("BIC score per model")
-xpos = (
-    np.mod(bic.argmin(), len(n_components_range))
-    + 0.65
-    + 0.2 * np.floor(bic.argmin() / len(n_components_range))
+# -------------------
+#
+# In the present case, the model with 2 components and full covariance
+# (which corresponds to the true generative model) has the lowest BIC
+# score is therefore selected.
+
+import pandas as pd
+
+df = pd.DataFrame(grid_search.cv_results_)[
+    ["param_n_components", "param_covariance_type", "mean_test_score"]
+]
+df["mean_test_score"] = -df["mean_test_score"]
+df = df.rename(
+    columns={
+        "param_n_components": "Number of components",
+        "param_covariance_type": "Type of covariance",
+        "mean_test_score": "BIC score",
+    }
 )
-plt.text(xpos, bic.min() * 0.97 + 0.03 * bic.max(), "*", fontsize=14)
-spl.set_xlabel("Number of components")
-spl.legend([b[0] for b in bars], cv_types)
-
-# Plot the winner
-splot = plt.subplot(2, 1, 2)
-Y_ = clf.predict(X)
-for i, (mean, cov, color) in enumerate(zip(clf.means_, clf.covariances_, color_iter)):
+df.head()
```

Comment:
Before showing the head, we should probably sort the raws based on the BIC score.
It would be more intuitive to observer the five best candidates.

Reply:
```suggestion
df.sort_values(by="BIC score").head()
```